src/librustc_mir/transform/deaggregator.rs

```diff
@@ -21,116 +21,103 @@ impl MirPass for Deaggregator {
                           tcx: TyCtxt<'a, 'tcx, 'tcx>,
                           source: MirSource,
                           mir: &mut Mir<'tcx>) {
-        let node_path = tcx.item_path_str(source.def_id);
-        debug!("running on: {:?}", node_path);
-        // we only run when mir_opt_level > 2
-        if tcx.sess.opts.debugging_opts.mir_opt_level <= 2 {
-            return;
-        }
-
         // Don't run on constant MIR, because trans might not be able to
         // evaluate the modified MIR.
         // FIXME(eddyb) Remove check after miri is merged.
         let id = tcx.hir.as_local_node_id(source.def_id).unwrap();
         match (tcx.hir.body_owner_kind(id), source.promoted) {
-            (hir::BodyOwnerKind::Fn, None) => {},
-            _ => return
+            (_, Some(_)) |
+            (hir::BodyOwnerKind::Const, _) |
+            (hir::BodyOwnerKind::Static(_), _) => return,
+
+            (hir::BodyOwnerKind::Fn, _) => {
+                if tcx.is_const_fn(source.def_id) {
+                    // Don't run on const functions, as, again, trans might not be able to evaluate
+                    // the optimized IR.
+                    return
+                }
+            }
         }
-        // In fact, we might not want to trigger in other cases.
-        // Ex: when we could use SROA.  See issue #35259
 
-        for bb in mir.basic_blocks_mut() {
-            let mut curr: usize = 0;
-            while let Some(idx) = get_aggregate_statement_index(curr, &bb.statements) {
-                // do the replacement
-                debug!("removing statement {:?}", idx);
-                let src_info = bb.statements[idx].source_info;
-                let suffix_stmts = bb.statements.split_off(idx+1);
+        let can_deaggregate = |statement: &Statement| {
+            if let StatementKind::Assign(_, ref rhs) = statement.kind {
+                if let Rvalue::Aggregate(..) = *rhs {
+                    return true;
+                }
+            }
+
+            false
+        };
+
+        let (basic_blocks, local_decls) = mir.basic_blocks_and_local_decls_mut();
+        for bb in basic_blocks {
+            let mut start = 0;
+            while let Some(i) = bb.statements[start..].iter().position(&can_deaggregate) {
+                let i = start + i;
+
+                // FIXME(eddyb) this is probably more expensive than it should be.
```

Comment:
Seems fine, you could also `drain()` and accumulate into a fresh vector or something... 

```rust
let mut new_statements = vec![]; 
for statement in bb.statements.drain(..) {
    if !can_deaggregate(statement) {
        new_statements.push(statement);
    }
    ...
}
```

though probably you would want to test first that there is at least one match.